### PR TITLE
SUS-2779 Replace in-cluster users table JOINs with external database 

### PR DESCRIPTION
**Pull request by v3rth**

### CODE REVIEW

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/72b0a80ae843baa2f534a49622242805063a0bf5/includes/api/ApiQueryLogEvents.php#L43): The Cyclomatic Complexity of this function "execute" is 23 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [LogPager.php#L29](https://github.com/Wikia/app/blob/72b0a80ae843baa2f534a49622242805063a0bf5/includes/logging/LogPager.php#L29): Class "LogPager" has 23 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: includes/cache/GenderCache.php [hunk 1]
```diff
@@ -110,26 +122,32 @@ public function doQuery( $users, $caller = '' ) {
...
+	private function getGenderOfUsersFromDB( $userNames, $caller ): ResultWrapper {
+		global $wgExternalSharedDB;
+		$dbr = wfGetDB( DB_SLAVE, [], $wgExternalSharedDB );
+
+		$table = [ 'user', 'user_properties' ];
+		$fields = [ 'user_name', 'up_value' ];
+		$conds = [ 'user_name' => $userNames ];
+		$joins = [
+			'user_properties' => [
+				'LEFT JOIN', [ 'user_id = up_user', 'up_property' => 'gender' ],
```
Comment (mszabo-wikia) [inline includes/cache/GenderCache.php:143]: I think we should be vary of joining against `wikicities.user` table as we do not know yet what will be the future single source of truth for user info.

**Review by mszabo-wikia — CHANGES_REQUESTED** [review]

File: includes/EditPage.php [hunk 2]
```diff
@@ -2621,26 +2621,26 @@ protected function wasDeletedSinceLastEdit() {
...
+		], [
+			'log_namespace' => $this->mTitle->getNamespace(),
+			'log_title' => $this->mTitle->getDBkey(),
+			'log_type' => 'delete',
+			'log_action' => 'delete',
+		], __METHOD__, [
+			'LIMIT' => 1,
+			'ORDER BY' => 'log_timestamp DESC'
+		] );
+
```
Comment (macbre) [inline includes/EditPage.php:2643]: `user_name` is no longer returned by this query. Is it calculated later on?

Comment (macbre) [reply includes/EditPage.php:2643]: Please mark core changes with `SUS-...` ticket ID. We do still perform backports from MediaWiki upstream. Such comments help a lot.

File: includes/Revision.php [hunk 3]
```diff
@@ -295,20 +295,14 @@ private static function loadFromConds( $db, $conditions, $flags = 0 ) {
...
-	public static function fetchRevision( $title ) {
-		return self::fetchFromConds(
-			wfGetDB( DB_SLAVE ),
-			array( 'rev_id=page_latest',
-				   'page_namespace' => $title->getNamespace(),
-				   'page_title'     => $title->getDBkey() )
-		);
+	private static function replaceUsernameFieldsWithVariables( $row )
+	{
+		$user = User::newFromId( $row->rev_user );
```
Comment (macbre) [inline includes/Revision.php]: `rev_user` can be zero. In such cases it's an anon edit and rev_user_text has IP address. We should use `User::getUsername` here instead that handles such cases.

File: includes/api/ApiQueryLogEvents.php [hunk 4]
```diff
@@ -157,13 +153,26 @@ public function execute() {
...
 				break;
 			}
 
+			if ( $this->fld_userid ) {
+				$row->user_id = $row->log_user;
+			}
+
+			if ( $this->fld_user ) {
+				$row->user_name = isset($userNames[$row->log_user]) ?
+					$userNames[$row->log_user] : 'unknown';
```
Comment (macbre) [inline includes/api/ApiQueryLogEvents.php]: I'd rather log a warning here. There should not be any mismatch of data. Returning `unknown` will not help in debugging :smiley:

File: includes/api/ApiQueryLogEvents.php [hunk 5]
```diff
@@ -469,4 +478,33 @@ public function getHelpUrls() {
...
+
+	private function getIndexedUserNames( $userIds ) {
+		global $wgExternalSharedDB;
+		if (count($userIds) === 0) {
+			return [];
+		}
+
+		$db = wfGetDB( DB_SLAVE, [], $wgExternalSharedDB );
+		$conditions = [ 'user_id' => array_unique( $userIds ) ];
+		$dbResults = $db->select( 'user', [ 'user_id', 'user_name' ], $conditions );
```
Comment (macbre) [inline includes/api/ApiQueryLogEvents.php]: Pass `__METHOD__` as `$fname` argument here.

File: includes/logging/LogPager.php [hunk 6]
```diff
@@ -296,7 +303,65 @@ public function getStartBody() {
...
+	 */
+	private function joinUserDataToLogRow( $row, $userNames ) {
+		$row->user_id = $row->log_user;
+
+		if ( isset( $userNames[$row->user_id] ) ) {
+			$row->user_name = $userNames[$row->user_id]->user_name;
+			$row->user_text = $row->user_name;
+			$row->user_editcount = $userNames[$row->user_id]->user_editcount;
+		} else {
+			$row->user_name = 'unknown';
```
Comment (macbre) [inline includes/logging/LogPager.php]: This should not happen. Let's log a warning here.

**Review by macbre — CHANGES_REQUESTED** [review]

File: includes/EditPage.php (hunk 2 above)
Comment (v3rth) [reply includes/EditPage.php:2643]: No it isn't calculated. This method is used only in [\EditPage::wasDeletedSinceLastEdit](https://github.com/Wikia/app/blob/72b0a80ae843baa2f534a49622242805063a0bf5/includes/EditPage.php#L2610) and there is no need to get any of user data in this case

Comment (mszabo-wikia) [reply includes/EditPage.php:2643]: But the user name of last delete is used on https://github.com/Wikia/app/blob/72b0a80ae843baa2f534a49622242805063a0bf5/includes/EditPage.php#L1841 .

Comment (v3rth) [reply includes/EditPage.php:2643]: Thanks! I didn't catch that one

File: includes/cache/GenderCache.php (hunk 1 above)
Comment (v3rth) [reply includes/cache/GenderCache.php:143]: So what should I use instead? Separated select and match results?

File: includes/api/ApiQueryLogEvents.php [hunk 7]
```diff
@@ -157,13 +154,33 @@ public function execute() {
...
+				$row->user_id = $row->log_user;
+			}
+
+			if ( $this->fld_user ) {
+				if ( isset( $userNames[$row->log_user] ) ) {
+					$row->user_name = $userNames[$row->log_user];
+				} else {
+					$row->user_name = 'unknown';
+					\Wikia\Logger\WikiaLogger::instance()
+						->warning( "User with id {$$row->log_user} was not found" );
```
Comment (WikiaTech) [inline includes/api/ApiQueryLogEvents.php]: MAJOR Remove the use of this variable variable "$$row". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1599)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/d020ff3eaf745a6017bee24eee73961c47368366/includes/api/ApiQueryLogEvents.php#L43): The Cyclomatic Complexity of this function "execute" is 23 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/d020ff3eaf745a6017bee24eee73961c47368366/includes/api/ApiQueryLogEvents.php#L43): This function "execute" has 153 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)
1. ![MAJOR][MAJOR] [LogPager.php#L29](https://github.com/Wikia/app/blob/d020ff3eaf745a6017bee24eee73961c47368366/includes/logging/LogPager.php#L29): Class "LogPager" has 23 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/52327bbc86529d1d76ca2b0b5f9fc5e9e26874fc/includes/api/ApiQueryLogEvents.php#L43): The Cyclomatic Complexity of this function "execute" is 23 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/52327bbc86529d1d76ca2b0b5f9fc5e9e26874fc/includes/api/ApiQueryLogEvents.php#L43): This function "execute" has 153 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)
1. ![MAJOR][MAJOR] [LogPager.php#L29](https://github.com/Wikia/app/blob/52327bbc86529d1d76ca2b0b5f9fc5e9e26874fc/includes/logging/LogPager.php#L29): Class "LogPager" has 23 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: includes/api/ApiQueryLogEvents.php [hunk 8]
```diff
@@ -169,8 +169,13 @@ public function execute() {
...
 
 			if ( $this->fld_user ) {
-				$row->user_name = isset($userNames[$row->log_user]) ?
-					$userNames[$row->log_user] : 'unknown';
+				if ( isset( $userNames[$row->log_user] ) ) {
+					$row->user_name = $userNames[$row->log_user];
+				} else {
+					$row->user_name = 'unknown';
+					\Wikia\Logger\WikiaLogger::instance()
+						->warning( "User with id {$$row->log_user} was not found" );
```
Comment (macbre) [inline includes/api/ApiQueryLogEvents.php]: Why `$$`? :)

Comment (macbre) [reply includes/api/ApiQueryLogEvents.php]: I'd rather use a fixed log message, e.g. `User not found by a given ID` and log the ID in the log message context.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/353b0871da57b009a8ed3c0fde81d1ddc53c2e89/includes/api/ApiQueryLogEvents.php#L43): The Cyclomatic Complexity of this function "execute" is 23 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [ApiQueryLogEvents.php#L43](https://github.com/Wikia/app/blob/353b0871da57b009a8ed3c0fde81d1ddc53c2e89/includes/api/ApiQueryLogEvents.php#L43): This function "execute" has 153 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)
1. ![MAJOR][MAJOR] [LogPager.php#L29](https://github.com/Wikia/app/blob/353b0871da57b009a8ed3c0fde81d1ddc53c2e89/includes/logging/LogPager.php#L29): Class "LogPager" has 23 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: includes/api/ApiQueryLogEvents.php (hunk 8 above)
Comment (v3rth) [reply includes/api/ApiQueryLogEvents.php]: `$$` was a typo, fixed in https://github.com/Wikia/app/pull/13912/commits/52327bbc86529d1d76ca2b0b5f9fc5e9e26874fc

File: includes/cache/GenderCache.php (hunk 1 above)
Comment (ghost) [reply includes/cache/GenderCache.php:143]: Let's keep it for now.  I will file a ticket for refining this after the initial release. https://wikia-inc.atlassian.net/browse/SUS-2979

**Review by ghost — APPROVED** [review]

**Review by mszabo-wikia — APPROVED** [review]
:shipit:

**Review by macbre — APPROVED** [review]

**Pull request merged**